File: cslacs/Templates/vb/EditableRootParent.vb
Imports System
Imports Csla
Imports Csla.Security

<Serializable()> _
Public Class EditableRootParent
  Inherits BusinessBase(Of EditableRootParent)
#Region "Business Methods"

  ' TODO: add your own fields, properties and methods 

  ' example with private backing field 
  Private Shared IdProperty As PropertyInfo(Of Integer) = RegisterProperty(New PropertyInfo(Of Integer)("Id"))
  Private _Id As Integer = IdProperty.DefaultValue
  Public Property Id() As Integer
    Get
      Return GetProperty(IdProperty, _Id)
    End Get
    Set(ByVal value As Integer)
      SetProperty(IdProperty, _Id, value)
    End Set
  End Property

  ' example with managed backing field 
  Private Shared NameProperty As PropertyInfo(Of String) = RegisterProperty(New PropertyInfo(Of String)("Name"))
  Public Property Name() As String
    Get
      Return GetProperty(NameProperty)
    End Get
    Set(ByVal value As String)
      SetProperty(NameProperty, value)
    End Set
  End Property

  Private Shared ChildListProperty As PropertyInfo(Of EditableChildList) = RegisterProperty(Of EditableChildList)(New PropertyInfo(Of EditableChildList)("ChildList", "Child list"))
  Public ReadOnly Property ChildList() As EditableChildList
    Get
      Return GetProperty(Of EditableChildList)(ChildListProperty)
    End Get
  End Property

  Private Shared ChildProperty As PropertyInfo(Of EditableChild) = RegisterProperty(New PropertyInfo(Of EditableChild)("Child", "Child"))
  Public ReadOnly Property Child() As EditableChild
    Get
      Return GetProperty(Of EditableChild)(ChildProperty)
    End Get
  End Property
#End Region

#Region "Validation Rules"

  Protected Overloads Overrides Sub AddBusinessRules()
    ' TODO: add validation rules 
    'ValidationRules.AddRule(RuleMethod, ""); 
  End Sub

#End Region

#Region "Authorization Rules"

  Protected Overloads Overrides Sub AddAuthorizationRules()
    ' TODO: add authorization rules 
    AuthorizationRules.AllowWrite(NameProperty, "Role")
  End Sub

  Private Shared Sub AddObjectAuthorizationRules()
    ' TODO: add authorization rules 
    'AuthorizationRules.AllowEdit(GetType(EditableRootParent), "Role")
  End Sub

#End Region

#Region "Factory Methods"

  Public Shared Function NewEditableRootParent() As EditableRootParent
    Return DataPortal.Create(Of EditableRootParent)()
  End Function

  Public Shared Function GetEditableRootParent(ByVal id As Integer) As EditableRootParent
    Return DataPortal.Fetch(Of EditableRootParent)(New SingleCriteria(Of EditableRootParent, Integer)(id))
  End Function

  Public Shared Sub DeleteEditableRootParent(ByVal id As Integer)
    DataPortal.Delete(New SingleCriteria(Of EditableRootParent, Integer)(id))
  End Sub

  ' Require use of factory methods 
  Private Sub New()
  End Sub

#End Region

#Region "Data Access"

  <RunLocal()> _
  Protected Overloads Overrides Sub DataPortal_Create()
    ' TODO: load default values 
    ' omit this override if you have no defaults to set 
    LoadProperty(ChildListProperty, EditableChildList.NewEditableChildList())
    LoadProperty(ChildProperty, EditableChild.NewEditableChild())
    MyBase.DataPortal_Create()
  End Sub

  Private Overloads Sub DataPortal_Fetch(ByVal criteria As SingleCriteria(Of EditableRootParent, Integer))
    ' TODO: load values 
    LoadProperty(ChildListProperty, EditableChildList.GetEditableChildList(Nothing))
    LoadProperty(ChildProperty, EditableChild.GetEditableChild(Nothing))
  End Sub

  <Transactional(TransactionalTypes.TransactionScope)> _
  Protected Overloads Overrides Sub DataPortal_Insert()
    ' TODO: insert values 
    FieldManager.UpdateChildren(Me)
  End Sub

  <Transactional(TransactionalTypes.TransactionScope)> _
  Protected Overloads Overrides Sub DataPortal_Update()
    ' TODO: update values 
    FieldManager.UpdateChildren(Me)
  End Sub

  <Transactional(TransactionalTypes.TransactionScope)> _
  Protected Overloads Overrides Sub DataPortal_DeleteSelf()
    DataPortal_Delete(New SingleCriteria(Of EditableRootParent, Integer)(Me.Id))
  End Sub

  <Transactional(TransactionalTypes.TransactionScope)> _
  Private Overloads Sub DataPortal_Delete(ByVal criteria As SingleCriteria(Of EditableRootParent, Integer))
    ' TODO: delete values 
    FieldManager.UpdateChildren(Me)
  End Sub

#End Region
End Class

File: Support/Templates/content/CslaLibrary-CSharp/EditableRootParent.cs
using System;
using Csla;

namespace Company.CslaLibrary1
{
  [Serializable]
  public class EditableRootParent : BusinessBase<EditableRootParent>
  {
    public static readonly PropertyInfo<int> IdProperty = RegisterProperty<int>(nameof(Id));
    public int Id
    {
      get { return GetProperty(IdProperty); }
      set { SetProperty(IdProperty, value); }
    }

    public static readonly PropertyInfo<string> NameProperty = RegisterProperty<string>(nameof(Name));
    public string Name
    {
      get { return GetProperty(NameProperty); }
      set { SetProperty(NameProperty, value); }
    }

    // example with Editable Child list 
    public static readonly PropertyInfo<EditableChildList> ChildListProperty = RegisterProperty<EditableChildList>(nameof(ChildList), "Child list");
    public EditableChildList ChildList
    {
      get { return GetProperty<EditableChildList>(ChildListProperty); }
    }

    // Example with Editable Child 
    public static readonly PropertyInfo<EditableChild> ChildProperty = RegisterProperty<EditableChild>(nameof(Child), "Child");
    public EditableChild Child
    {
      get { return GetProperty<EditableChild>(ChildProperty); }
    }

    protected override void AddBusinessRules()
    {
      // TODO: add validation rules
      //BusinessRules.AddRule(new Rule(), IdProperty);
    }

    private static void AddObjectAuthorizationRules()
    {
      // TODO: add authorization rules
      //BusinessRules.AddRule(...);
    }

    [RunLocal]
    [Create]
    private void Create()
    {
      // TODO: load default values
      // omit this override if you have no defaults to set
      var listDataPortal = ApplicationContext.GetRequiredService<IChildDataPortal<EditableChildList>>();
      LoadProperty(ChildListProperty, listDataPortal.CreateChild());
      var itemDataPortal = ApplicationContext.GetRequiredService<IChildDataPortal<EditableChild>>();
      LoadProperty(ChildProperty, itemDataPortal.CreateChild());
      BusinessRules.CheckRules();
    }

    [Fetch]
    private void Fetch(int id)
    {
      // TODO: load values
      var listDataPortal = ApplicationContext.GetRequiredService<IChildDataPortal<EditableChildList>>();
      LoadProperty(ChildListProperty, listDataPortal.FetchChild(null));
      var itemDataPortal = ApplicationContext.GetRequiredService<IChildDataPortal<EditableChild>>();
      LoadProperty(ChildProperty, itemDataPortal.FetchChild(null));
    }

    [Insert]
    [Transactional(TransactionalTypes.TransactionScope)]
    private void Insert()
    {
      // TODO: insert values
      FieldManager.UpdateChildren(this);
    }

    [Update]
    [Transactional(TransactionalTypes.TransactionScope)]
    private void Update()
    {
      // TODO: update values
      FieldManager.UpdateChildren(this);
    }

    [DeleteSelf]
    [Transactional(TransactionalTypes.TransactionScope)]
    private void DeleteSelf()
    {
      Delete(this.Id);
    }

    [Delete]
    [Transactional(TransactionalTypes.TransactionScope)]
    private void Delete(int id)
    {
      // TODO: delete values
      FieldManager.UpdateChildren(this);
    }
  }
}
